packages/near-membrane-base/src/membrane.ts

```diff
@@ -1167,166 +885,322 @@ export function createMembraneMarshall() {
+                : BoundaryProxyHandler.passthruGetTrap;
+
+            private static defaultHasTrap = trapMutations
+                ? BoundaryProxyHandler.hybridHasTrap
+                : BoundaryProxyHandler.passthruHasTrap;
+
+            private static defaultIsExtensibleTrap = BoundaryProxyHandler.passthruIsExtensibleTrap;
+
+            private static defaultOwnKeysTrap = BoundaryProxyHandler.passthruOwnKeysTrap;
+
+            private static defaultPreventExtensionsTrap = trapMutations
+                ? BoundaryProxyHandler.pendingPreventExtensionsTrap
+                : BoundaryProxyHandler.passthruPreventExtensionsTrap;
+
+            private static defaultSetTrap = trapMutations
+                ? BoundaryProxyHandler.pendingSetTrap
+                : BoundaryProxyHandler.passthruSetTrap;
+
+            private static defaultSetPrototypeOfTrap = trapMutations
+                ? BoundaryProxyHandler.pendingSetPrototypeOfTrap
+                : BoundaryProxyHandler.passthruSetPrototypeOfTrap;
+
+            // hybrid traps
+            // (traps that operate on their shadowTarget, proxy, and foreignTargetPointer):
+
+            private static hybridGetTrap(
+                this: BoundaryProxyHandler,
+                shadowTarget: ShadowTarget,
+                key: string | symbol,
+                receiver: any
+            ): ReturnType<typeof Reflect.get> {
+                const { foreignTargetPointer } = this;
+                const safeDesc =
+                    getOwnForeignDescriptor(foreignTargetPointer, shadowTarget, key) ||
+                    getInheritedForeignDescriptor(foreignTargetPointer, key);
+
+                if (safeDesc) {
+                    const { get: getter, value: localValue } = safeDesc;
+                    if (getter) {
+                        // Even though the getter function exists, we can't use
+                        // `ReflectGet` because there might be a distortion for
+                        // that getter function, in which case we must resolve
+                        // the local getter and call it instead.
+                        return ReflectApply(getter, receiver, []);
+                    }
+                    return localValue;
+                }
+                if (key === TO_STRING_TAG_SYMBOL) {
+                    return foreignCallableGetUnbrandedTag(foreignTargetPointer);
+                }
+                return undefined;
+            }
+
+            private static hybridHasTrap(
+                this: BoundaryProxyHandler,
+                _shadowTarget: ShadowTarget,
+                key: string | symbol
+            ): ReturnType<typeof Reflect.has> {
+                const { foreignTargetPointer } = this;
+                if (foreignCallableHasOwnProperty(foreignTargetPointer, key)) {
+                    return true;
+                }
+                // Avoiding calling the has trap for any proto chain operation,
```

Comment:
this block should probably we abstracted into something equivalent to `getInheritedForeignDescriptor`

Reply:
They are very similar. If it's used more than once then okay. I want to avoid unneeded abstraction in this low level stuff.